pyomo/contrib/sensitivity_toolbox/sens.py

```diff
@@ -7,216 +7,684 @@
-        m = instance
-    
-    #Generate component maps for associating Variables to perturbations
-    varSubList = []
-    for parameter in paramSubList:
-        tempName = unique_component_name(b,parameter.local_name)
-        b.add_component(tempName,Var(parameter.index_set()))
-        myVar = b.component(tempName)
-        varSubList.append(myVar)
- 
-    #Note: substitutions are not currently compatible with 
-    #      ComponentMap [ECSA 2018/11/23], this relates to Issue #755
-    paramCompMap = ComponentMap(zip(paramSubList, varSubList))
-    variableSubMap = {}
-    #variableSubMap = ComponentMap()
-    paramPerturbMap = ComponentMap(zip(paramSubList,perturbList))
-    perturbSubMap = {}
-    #perturbSubMap = ComponentMap()
-   
-    paramDataList = [] 
-    for parameter in paramSubList:
-        # Loop over each ParamData in the Param Component
-        #
-        # Note: Sets are unordered in Pyomo.  For this to be
-        # deterministic, we need to sort the index (otherwise, the
-        # ordering of things in the paramDataList may change).  We use
-        # sorted_robust to guard against mixed-type Sets in Python 3.x
-        for kk in sorted_robust(parameter):
-            variableSubMap[id(parameter[kk])]=paramCompMap[parameter][kk]
-            perturbSubMap[id(parameter[kk])]=paramPerturbMap[parameter][kk]
-            paramDataList.append(parameter[kk])
-
-    #clone Objective, add to Block, and update any Expressions
-    for cc in list(m.component_data_objects(Objective,
-                                            active=True,
-                                            descend_into=True)):
-        tempName=unique_component_name(m,cc.local_name)    
-        b.add_component(tempName,
-                  Objective(expr=ExpressionReplacementVisitor(
-                  substitute=variableSubMap,
-                  remove_named_expressions=True).dfs_postorder_stack(cc.expr)))
-        cc.deactivate()
-    
-    #clone Constraints, add to Block, and update any Expressions
-    b.constList = ConstraintList()
-    for cc in list(m.component_data_objects(Constraint, 
-                                   active=True,
-                                   descend_into=True)):
-        if cc.equality:
-            b.constList.add(expr= ExpressionReplacementVisitor(
-                    substitute=variableSubMap,
-                    remove_named_expressions=True).dfs_postorder_stack(cc.expr))
+        gradient_c = np.array([])
+        gradient_c_dic = {}
+    # remove all generated files
+    shutil.move("col_row.nl", "./GJH/")
+    shutil.move("col_row.col", "./GJH/")
+    shutil.move("col_row.row", "./GJH/")
+    shutil.rmtree('GJH', ignore_errors=True)
+    return gradient_f,gradient_f_dic, gradient_c,gradient_c_dic, line_dic
+
+def line_num(file_name, target):
+    """This function returns the line number contains 'target' in the file_name.
+    This function identities constraints that have variables in theta_names.
```

Comment:
How does this function identify constraints that have variables in theta_names? Is this line accurate?

Reply:
It searches variables in the input file. I added two unittests for this. 


    def test_line_num1(self):
          import os
          file_name = "test_col.col"
          with open(file_name, "w") as file:
              file.write("var1\n")
              file.write("var3\n")
          i= line_num(file_name,'var1')
          j= line_num(file_name,'var3')
          self.assertEqual(i, 1)
          self.assertEqual(j, 2)

    def test_line_num1(self):
        import os
        file_name = "test_col.col"
        with open(file_name, "w") as file:
            file.write("var1\n")
            file.write("var3\n")
        with self.assertRaises(Exception) as context:
            i= line_num(file_name,'var2')
        self.assertTrue('test_col.col does not include var2' in str(context.exception))